linenoise/linenoise/linenoise.h

```diff
@@ -0,0 +1,335 @@
+
+/**
+ * @brief Creates a new linenoise instance.
+ *
+ * @param param Pointer to the configuration parameters for the instance.
+ * @return Handle to the created instance, or NULL on failure.
+ */
+linenoise_handle_t linenoise_create_instance(linenoise_instance_param_t *param);
+
+/**
+ * @brief Destroys a linenoise instance and frees associated memory.
+ *
+ * @param handle Handle to the instance to delete.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_delete_instance(linenoise_handle_t handle);
+
+/**
+ * @brief Reads a line of input from the user into a buffer.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @param cmd_line_buffer Buffer to store the input line.
+ * @param cmd_line_length Length of the command line buffer.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_get_line(linenoise_handle_t handle, char *cmd_line_buffer, size_t cmd_line_length);
+
+/**
+ * @brief Adds a line to the instance's history.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @param line The line to add to history.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_history_add(linenoise_handle_t handle, const char *line);
+
+/**
+ * @brief Saves the history to a file.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @param filename Path to the history file.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_history_save(linenoise_handle_t handle, const char *filename);
+
+/**
+ * @brief Loads history from a file.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @param filename Path to the history file.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_history_load(linenoise_handle_t handle, const char *filename);
+
+/**
+ * @brief Sets the maximum number of entries in the history.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @param len Maximum number of entries.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_history_set_max_len(linenoise_handle_t handle, int len);
+
+/**
+ * @brief Frees the history associated with the instance.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_history_free(linenoise_handle_t handle);
+
+/**
+ * @brief Clears the terminal screen for the instance.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_clear_screen(linenoise_handle_t handle);
+
+/**
+ * @brief Sets whether an empty line is allowed to be returned.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @param empty_line true to allow empty lines, false to disallow.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t linenoise_set_empty_line(linenoise_handle_t handle, bool empty_line);
```

Comment:
Is there a use case for changing these settings (empty_line, multi_line, dumb_mode, max_cmd_line_length) once the instance is already created? If there is no concrete use case, it might be simpler to not implement these getters and setters.

Reply:
If the user creates a linenoise instance and uses it for a given console instance, then if the user wants to delete the console instance used with the linenoise instance and create a new console instance with different linenoise specs, it could be worth keeping those getters and setters. Although if we follow through with this idea we might want to add other getters and setters to be able to change other config fields (e.g., prompt).

Although deleting and creating a new line oise instance will do the job.